File: 프로그래머스/lv3/43162. 네트워크/네트워크.java
class Solution {
    public int solution(int n, int[][] computers) {
        int answer = 0;
        boolean[] visited = new boolean[computers.length];
        
        for(int i = 0; i < computers.length; i++){
            visited[i] = false;
        }
        
        for(int i =0; i < computers.length; i++){
            if(visited[i] == false) {
                answer ++;
                dfs(i, visited, computers);
            }
        }
        
        return answer;
    }
    
    void dfs(int com, boolean[] visited, int[][] computers){
        visited[com] = true; // 방문 true
        
        for(int i = 0; i < computers.length; i++){
            if(visited[i] == false && computers[com][i]==1){
                dfs(i, visited, computers); //방문안하고 1이면
            }
        }
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

def solution(n, computers):
    
    graph = [[] for _ in range(n)]
    
    for i in range(n):
        for j in range(len(computers[i])):
            if computers[i][j] == 1 and i != j: # i, j 가 연결되어 있음
                graph[i].append(j)
    
    print(graph)
    
    # 전체 다 방문전으로 처리
    visited = [False] * n
    
    q = deque()

    def bfs(start):
        q.append(start)
        visited[start] = True
        while q:
            now = q.popleft()
            for nxt in graph[now]:
                if not visited[nxt]:
                    visited[nxt] = True
                    q.append(nxt)
        
    count = 0
    
    for i in range(n):
        if not visited[i]:
            bfs(i)
            count += 1
   
    print(count)
    return count
            
        
